File: src/main/java/frc/team5190/robot/navigation/NAVHelper.java
package frc.team5190.robot.navigation;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.util.StringTokenizer;

public class NAVHelper {

    private static final int NUM_ROWS = 54;

    public static double[][] leftPoints = new double[NUM_ROWS][3];
    public static double[][] rightPoints = new double[NUM_ROWS][3];

    public static void configPointsFromCSV(AutoMode mode) throws Exception {

        String leftPath;
        String rightPath;

        switch (mode) {
            case LEFT:
                // TODO
                leftPath = "C:\\Users\\prate\\Downloads\\testpath_left.csv";
                rightPath = "C:\\Users\\prate\\Downloads\\testpath_right.csv";
                break;

            case CENTER:
                leftPath = "C:\\Users\\prate\\Downloads\\testpath_left.csv";
                rightPath = "C:\\Users\\prate\\Downloads\\testpath_right.csv";
                break;

            case RIGHT:
                // TODO
                leftPath = "C:\\Users\\prate\\Downloads\\testpath_left.csv";
                rightPath = "C:\\Users\\prate\\Downloads\\testpath_right.csv";
                break;

            default:
                leftPath = "C:\\Users\\prate\\Downloads\\testpath_left.csv";
                rightPath = "C:\\Users\\prate\\Downloads\\testpath_right.csv";
                break;
        }
        File left = new File(leftPath);
        File right = new File(rightPath);

        int row = 0;
        int col = 0;

        BufferedReader bufferedReader = new BufferedReader(new FileReader(left));
        String line;

        while ((line = bufferedReader.readLine()) != null && row < NUM_ROWS) {
            StringTokenizer st = new StringTokenizer(line, ", ");
            while (st.hasMoreTokens()) {
                leftPoints[row][col] = Double.parseDouble(st.nextToken());
                col++;
            }
            col = 0;
            row++;
        }

        bufferedReader = new BufferedReader(new FileReader(right));
        row = 0;
        col = 0;

        while ((line = bufferedReader.readLine()) != null & row < NUM_ROWS) {
            StringTokenizer st = new StringTokenizer(line, ",");
            while (st.hasMoreTokens()) {
                rightPoints[row][col] = Double.parseDouble(st.nextToken());
                col++;
            }
            col = 0;
            row++;
        }
    }

    public enum AutoMode {
        LEFT, CENTER, RIGHT;
    }
}

File: src/main/kotlin/frc/team5190/robot/navigation/NAVHelper.kt
package frc.team5190.robot.navigation

import frc.team5190.robot.util.Hardware
import frc.team5190.robot.util.Maths
import java.io.InputStreamReader


enum class NAVHelper(private val leftFilePath: String, private val rightFilePath: String) {

    LEFTS_LEFT("left/left_left.csv", "left/left_right.csv"),
    LEFTS_RIGHT("left/right_let.csv", "left/right_right.csv"),
    CENTERS_LEFT("center/left_left.csv", "center/left_right.csv"),
    CENTERS_RIGHT("center/right_left.csv", "center/right_right.csv"),
    RIGHTS_LEFT("right/left_left.csv", "right/left_right.csv"),
    RIGHTS_RIGHT("right/right_left.csv", "right/right_right.csv");

    val trajectoryLeft by lazy {
        javaClass.classLoader.getResourceAsStream(leftFilePath).use { stream ->
            return@lazy InputStreamReader(stream).readLines().map {
                val pointData = it.split(",").map { it.trim() }
                return@map TrajectoryData(pointData[0].toDouble(), pointData[1].toDouble(), pointData[2].toLong())
            }
        }
    }

    val trajectoryRight by lazy {
        javaClass.classLoader.getResourceAsStream(rightFilePath).use { stream ->
            return@lazy InputStreamReader(stream).readLines().map {
                val pointData = it.split(",").map { it.trim() }
                return@map TrajectoryData(pointData[0].toDouble(), pointData[1].toDouble(), pointData[2].toLong())
            }
        }
    }
}

typealias TrajectoryList = List<TrajectoryData>

data class TrajectoryData(private val position: Double, private val velocity: Double, val duration: Long) {
    val rotations = Maths.feetToRotations(position, Hardware.WHEEL_RADIUS.toDouble())
    val rpm = Maths.feetPerSecondToRPM(velocity, Hardware.WHEEL_RADIUS.toDouble())

    val nativeUnits = Maths.rotationsToNativeUnits(rotations, 1440.0)
    val nativeUnitsPer100Ms = Maths.rpmToNativeUnitsPer100Ms(rpm, 1440.0)
}


